src/shadowcasting.cpp

```diff
@@ -0,0 +1,663 @@
+
+                    if( !floor_block ) {
+                        ( *output_caches[z_index] )[current.x][current.y] =
+                            std::max( ( *output_caches[z_index] )[current.x][current.y], last_intensity );
+                    }
+
+                    if( !started_span ) {
+                        // Need to reset minor slope, because we're starting a new line
+                        new_start_minor = leading_edge_minor;
+                        started_span = true;
+                    }
+
+                    if( new_transparency == current_transparency ) {
+                        // All in order, no need to split the span.
+                        new_start_minor = leading_edge_minor;
+                        continue;
+                    }
+
+                    // Handle spliting the span into up to 4 separate spans
+                    split_span<T, is_transparent, accumulate>( spans, this_span, current_transparency,
+                            new_transparency, last_intensity,
+                            distance, new_start_minor,
+                            trailing_edge_major, leading_edge_major,
+                            trailing_edge_minor, leading_edge_minor );
+                }
+
+                // If we end the row with an opaque tile, set the span to start at the next row
+                // since we don't need to process the current one any more.
+                if( !is_transparent( current_transparency, last_intensity ) ) {
+                    this_span->start_major = leading_edge_major;
+                }
+            }
+
+            if( !started_block ) {
+                // If we didn't scan at least 1 z-level, don't iterate further
+                // Otherwise we may "phase" through tiles without checking them or waste time
+                // checking spans that are out of bounds.
+                this_span = spans.erase( this_span );
+            } else if( !is_transparent( current_transparency, last_intensity ) ) {
+                // If we reach the end of the span with terrain being opaque, we don't iterate further.
+                // This means that any encountered transparent tiles from the current span have been
+                // split off into new spans
+                this_span = spans.erase( this_span );
+            } else {
+                // Cumulative average of the values encountered.
+                this_span->cumulative_value = accumulate( this_span->cumulative_value,
+                                              current_transparency, distance );
+                ++this_span;
+            }
+        }
+    }
+}
+
+template<int x_transform, int y_transform, int z_transform, typename T,
+         T( *calc )( const T &, const T &, const int & ),
+         bool( *is_transparent )( const T &, const T & ),
+         T( *accumulate )( const T &, const T &, const int & )>
+void cast_vertical_zlight_segment(
```

Comment:
Similar comments apply to this function as the above.

Reply:
I think I applied the ones I could.